### ISSUE TITLE
[MU4 Issue] Playback curtailed on unison between voices

### ISSUE THREAD

**Issue by la-bel** [musescore/MuseScore#13853]
**Describe the bug**
When two voices are playing the same note for different durations, the longer playback is cut short to the length of the shorter playback.

**To Reproduce**
1. Create a new score with a sustaining instrument (like a saxophone)
2. Add a whole note in voice 1
3. Add a much shorter note in voice 2 starting on the same beat and playing the same note
4. Play the score - the longer note does not sustain. 

**Expected behavior**

The longer note should play for its written length.

**Screenshots**
<img width="964" alt="Screenshot 2022-10-20 at 21 43 07" src="https://user-images.githubusercontent.com/63920211/197054011-f2159fad-0ce6-44a7-abd9-8fcbca58bb4d.png">
Example

**Platform information**
 - OS: macOS

**Comment by MarcSabatella** [comment]
Not good indeed!  This was an ongoing issue with lots of debate over the years in previous versions of MuseScore as to what the desired behavior actually is with respect to the MIDI playback, and what the best way to achieve that behavior is.  The current implementation in MuseScore 3 is the best compromise people were able to come up with, and probably makes a good model.  See https://musescore.org/en/node/12971

**Comment by la-bel** [comment]
Thanks for the context! I couldn't think of ways to phrase it to see if there had been previous discussion.

**Comment by MarcSabatella** [comment]
@Tantacrul this is very serious

**Comment by Tantacrul** [comment]
Seems like something that can easily be solved now that we're using a new system.

Desired behaviour: the long and short duration in both voices should both be heard to their full extent.

**Comment by maestroharmony** [comment]
@Tantacrul sorry the ping but wouldn't this trigger Phasing issue like the same sample? I always thought that the reason this couldn't be achievable is because It's the same sample being triggered. I would suggest WHEN this happens, there would be a system that could trigger samples from a half/whole step down to use a different sample, otherwise it will become out-of-phase.

**Comment by Tantacrul** [comment]
> @Tantacrul sorry the ping but wouldn't this trigger a phasing issue like the same sample? I always thought that the reason this couldn't be achievable is because It's the same sample being triggered. I would suggest WHEN this happens, there would be a system that could trigger samples from a half/whole step down/up to use a different sample, otherwise it will become out-of-phase.

Let us worry about that.

**Comment by MarcSabatella** [comment]
Note there is also an issue even if the notes don't start at the same time, but it's different depending on MS Basic or Muse Sounds, also depending on whether it's all one one staff or two staves of a grand staff.

For instance, in the following example, MS Basic cuts the whole note short when the quarter note happens, and the tied whole note in the next measure is ignored.  Muse Sounds does not play the quarter at all, but repeats the tied note.  MuseScore 3 does what I'd consider the correct thing: the whole note sustains for eight beats but we hear a "restrike" of the note on beat 2.

![image](https://user-images.githubusercontent.com/1799936/202794810-79f7010e-9f8b-4b8b-83d6-6a458ad1f874.png)

Real world case in point: with Muse Sounds, the A4, D5, E5, and A5 don't play in the final cross-staff flourish in Reunion, but the final tied chord *does* get repeated, which it shouldn't (and doesn't if I remove the cross-staff notes.

![image](https://user-images.githubusercontent.com/1799936/202792907-d7ae60f8-bb28-44f6-8c65-6c38515999cf.png)

https://musescore.com/user/2975/scores/8899170/s/uaX1Pt?share=copy_link

**Comment by Tantacrul** [comment]
Moving to Post Release because it requires a significant change. Also, this doesn't appear to affect Muse Sounds, which mercifully makes it less bad

**Comment by MarcSabatella** [comment]
It *definitely* affects Muse Sounds, and in a way that is much worse for some scores - the subsequent notes don't play at all, and also tied notes get repeated as if the tie wasn't present.  See my example [above](https://github.com/musescore/MuseScore/issues/13853#issuecomment-1320483151), showing the simple case with the tied whole note, also with the ending of Reunion (which is the type of case for which Muse Sounds is very much worse than MS Basic).

**Comment by Tantacrul** [comment]
Ah, yes. My mistake about Muse Sounds. However, we're bumping the issue of unisons only. Any other issues to do with ties are more likely to be dealt with here.

**Comment by cbjeukendrup** [comment]
Here is a funny, very naive solution for FluidSynth: https://github.com/musescore/MuseScore/commit/a3e51b5c4f108b896429e55e1a3badef51ea4125
(with this solution, you won't hear a second attack for the second note though).
Not sure whether this solution deserves serious consideration, but it was a nice experiment anyway.

**Comment by MarcSabatella** [comment]
I don't get the joke :-), but not hearing the second attack isn't very reasonable if the second note happens later as it does in my examples above.  These are quite common situations in keyboard music, guitar music, and orchestra music (with shared parts on a single staff).

Anyone wanting to fix this might want to check out an earlier comment I made above, which is marked outdated but isn't - it's all still very relevant:

https://github.com/musescore/MuseScore/issues/13853#issuecomment-1286162537

Basically, I'm showing the earlier discussions that led to the solution used in MU3.  In those discussions you can see the code itself as well as the reason why that approach was chosen over other possibilities - reasons that I think still apply today.  There's more to it than the one thread I linked to, but you can follow the additional discussions by expanding out from there and following the links within that thread etc.

**Comment by Tantacrul** [comment]
Yes. When the same note is played by different voices, it is working.

When the same voice hits the same note, the 'cut off' problem still occurs (Organ is a good instrument to test) but I think this is much less serious.
![image](https://user-images.githubusercontent.com/21022311/205027025-2280ced4-dc36-42fe-9960-9cc7650a2cfa.png)

**Comment by MarcSabatella** [comment]
Oops, yes, somehow I managed to not test that case.

On piano, this is arguably the *correct* behavior anyhow, since there aren't two manuals that would allow one note to sustain after the other is played and released.  You'd need to rely on pedal to sustain the first note, and that works fine.  It's the multiple voices on one staff case that was more likely to be a problem in practice for other instruments, and that's fixed.  So I agree, this is probably good enough for now.

Interestingly, for the two staff case, simply moving one of those two notes to a different voice - while still keeping them on separate staves - produces the correct organ behavior.  Which is good to know just in case.  But it also makes me suspect this would actually be a trivial fix - the hard part is clearly done already.

**Comment by MarcSabatella** [comment]
This was never fixed for Muse Sounds it seems.

**Comment by rgreen5** [comment]
OS: Linux Mint 20.1, Arch.: x86_64, MuseScore version (64-bit): 4.1.0-231830502, revision: github-musescore-musescore-89b5a47

I think this bug may be related. Open the attached file and play measure 6. The half notes in voice 2 are abruptly cut off.
![notes_cut_off](https://github.com/musescore/MuseScore/assets/87246583/6b5770f5-c7a6-4ce6-9a9e-e9c4ab91dfd9)

[notes_cut_off.zip](https://github.com/musescore/MuseScore/files/11929588/notes_cut_off.zip)

**Comment by Jap2-0** [comment]
> Yes. When the same note is played by different voices, it is working.
> 
> When the same voice hits the same note, the 'cut off' problem still occurs (Organ is a good instrument to test) but I think this is much less serious. It is quite difficult to tell with piano because of the natural falling off of each note.
> 
> ![image](https://user-images.githubusercontent.com/21022311/205027025-2280ced4-dc36-42fe-9960-9cc7650a2cfa.png)
> 
> I'd say we're good to merge this and potentially deal with the 'two notes in voice 1' thing in a later release.

Is there an open issue for this, or should I create one?  I've been running into this a lot with an organ score I'm working on at the moment.

(Putting the other line in a different voice does fix playback, but leaves me with a voice full of rests.)
